platform/cxxsupport/mstd_mutex

```diff
@@ -0,0 +1,425 @@
+    scoped_lock &operator=(const scoped_lock &) = delete;
+}
+#else
+;
+#endif
+
+// 0 locks - no-op
+template <>
+class scoped_lock<> {
+public:
+    explicit scoped_lock() = default;
+    explicit scoped_lock(adopt_lock_t) noexcept { }
+    ~scoped_lock() = default;
+
+    scoped_lock(const scoped_lock &) = delete;
+    scoped_lock &operator=(const scoped_lock &) = delete;
+};
+
+// 1 lock - simple lock, equivalent to lock_guard<Mutex>
+template <class Mutex>
+class scoped_lock<Mutex> {
+    Mutex &pm;
+public:
+    using mutex_type = Mutex;
+    explicit scoped_lock(Mutex &m) : pm(m) { m.lock(); }
+    explicit scoped_lock(adopt_lock_t, Mutex &m) noexcept : pm(m) { }
+    ~scoped_lock() { pm.unlock(); }
+
+    scoped_lock(const scoped_lock &) = delete;
+    scoped_lock &operator=(const scoped_lock &) = delete;
+};
+#endif
+
+// [thread.once.onceflag]
+// Always local implementation - need to investigate GCC + ARMC6 retargetting
+struct once_flag {
+    constexpr once_flag() noexcept : __guard() { }
+    once_flag(const once_flag &) = delete;
+    once_flag &operator=(const once_flag &) = delete;
+    ~once_flag() = default;
+private:
+    template <class Callable, class... Args>
+    friend void call_once(once_flag &flag, Callable&& f, Args&&... args);
+    int __guard;
+};
+
+// [thread.once.callonce]
+template <class Callable, class... Args>
+void call_once(once_flag &flag, Callable&& f, Args&&... args)
+{
+    if (!(core_util_atomic_load_explicit(&flag.__guard, mbed_memory_order_acquire) & 1)) {
+        if (__cxa_guard_acquire(&flag.__guard)) {
+             mstd::invoke(mstd::forward<Callable>(f), mstd::forward<Args>(args)...);
+             __cxa_guard_release(&flag.__guard);
+        }
+    }
+}
+
+// [thread.mutex.class]
+// Always local implementation - need to investigate GCC + ARMC6 retargetting
+#if MBED_CONF_RTOS_PRESENT
+class _Mutex_base {
+    // Constructor must be constexpr - we are required to initialise on first use
+    // not in our constructor. (So that mutex use in static constructors is safe).
+    SingletonPtr<rtos::Mutex> _pm;
+public:
+    constexpr _Mutex_base() noexcept = default;
+    ~_Mutex_base();
+    _Mutex_base(const _Mutex_base &) = delete;
+    _Mutex_base &operator=(const _Mutex_base &) = delete;
+    void lock();
+    bool try_lock();
+    void unlock();
+};
+#else
+class _Mutex_base {
+public:
+    constexpr _Mutex_base() noexcept = default;
+    ~_Mutex_base() = default;
+    _Mutex_base(const _Mutex_base &) = delete;
+    _Mutex_base &operator=(const _Mutex_base &) = delete;
+    void lock() { }
+    bool try_lock() { return true; }
+    void unlock() { }
+};
+#endif
+
+// We don't currently distinguish implementations (and aren't required to -
+// current thread not owning a non-recursive one is a precondition, we don't
+// have to take any special action).
+class mutex : public _Mutex_base {
+};
+
+// [thread.mutex.recursive]
+class recursive_mutex : public _Mutex_base {
```

Comment:
any plan to add the `timed_` versions ?

Reply:
We need `chrono` for ARMC5 first. That feels like a separate project.

If licensing permitted just lifting the ARMC6 implementation...?

Reply:
This is one case where ARMC5 is a significant blocker. It would be easy to bring chrono in to use for the other 3 toolchains, but any attempt to keep ARMC5 building blocks it. Could make all relevant APIs conditional on `ifndef __CC_ARM`, but don't fancy that either.